core/services/workflows/workflow.go

```diff
@@ -97,20 +100,170 @@ type stepDefinition struct {
+func (s stepDefinitionYaml) toStepDefinition() stepDefinition {
+	return stepDefinition{
+		Ref:    s.Ref,
+		Type:   s.Type.String(),
+		Inputs: s.Inputs,
+		Config: s.Config,
+	}
+}
+
+type stepDefinitionType struct {
+	typeStr   string
+	typeTable *stepDefinitionTableType
+}
+
+func (s stepDefinitionType) String() string {
+	if s.typeStr != "" {
+		return s.typeStr
+	}
+
+	return s.typeTable.String()
+}
+
+func (s *stepDefinitionType) UnmarshalJSON(data []byte) error {
+	// Unmarshal the JSON data into a map to determine if it's a string or a table
+	var m string
+	err := json.Unmarshal(data, &m)
+	if err == nil {
+		s.typeStr = m
+		return nil
+	}
+
+	// If the JSON data is a table, unmarshal it into a stepDefinitionTableType
+	var table stepDefinitionTableType
+	err = json.Unmarshal(data, &table)
+	if err != nil {
+		return err
+	}
+	s.typeTable = &table
+	return nil
+}
+
+func (s *stepDefinitionType) MarshalJSON() ([]byte, error) {
+	if s.typeStr != "" {
+		return json.Marshal(s.typeStr)
+	}
+
+	return json.Marshal(s.typeTable)
+}
+
+// JSONSchema returns the JSON schema for a stepDefinitionType.
+//
+// The schema is a oneOf schema that allows either a string or a table.
+func (stepDefinitionType) JSONSchema() *jsonschema.Schema {
+	reflector := jsonschema.Reflector{DoNotReference: true, ExpandedStruct: true}
+	tableSchema := reflector.Reflect(&stepDefinitionTableType{})
+	stringSchema := &jsonschema.Schema{
+		Type:    "string",
+		Pattern: "^[a-z0-9_\\-:]+@(0|[1-9]\\d*)(?:-((?:0|[1-9]\\d*|\\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\\.(?:0|[1-9]\\d*|\\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\\+([0-9a-zA-Z-]+(?:\\.[0-9a-zA-Z-]+)*))?$",
+	}
+
+	return &jsonschema.Schema{
+		Title: "type",
+		OneOf: []*jsonschema.Schema{
+			stringSchema,
+			tableSchema,
+		},
+	}
+}
+
+// stepDefinitionTableType is the structured representation of a stepDefinitionType.
+type stepDefinitionTableType struct {
+	Name    string            `json:"name"`
+	Version string            `json:"version" jsonschema:"pattern=(0|[1-9]\\d*)(?:-((?:0|[1-9]\\d*|\\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\\.(?:0|[1-9]\\d*|\\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\\+([0-9a-zA-Z-]+(?:\\.[0-9a-zA-Z-]+)*))?$"`
+	Tags    map[string]string `json:"tags"`
+}
+
+// String returns the string representation of a stepDefinitionTableType.
+//
+// It follows the format:
+//
+//	{name}:{tag1_key}_{tag1_value}:{tag2_key}_{tag2_value}@{version}
+//
+// where tags are ordered alphanumerically.
+func (s stepDefinitionTableType) String() string {
+	tags := make([]string, 0, len(s.Tags))
+	for k, v := range s.Tags {
+		tags = append(tags, fmt.Sprintf("%s_%s", k, v))
+	}
+	slices.Sort(tags)
+
+	return fmt.Sprintf("%s:%s@%s", s.Name, strings.Join(tags, ":"), s.Version)
+}
+
+// workflowSpecYaml is the YAML representation of a workflow spec.
+//
+// It allows for multiple ways of defining a workflow spec, which we later
+// convert to a single representation, `workflowSpec`.
+type workflowSpecYaml struct {
```

Comment:
As an organization note: it might be worth moving the models to separate files, eg. models.go for the format-agnostic models and models_yaml for the yaml specific ones.